Added better tests for math ops and unified them

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21131 Added degrees, radians, ldexp
* #21130 Added frexp, isinf, isnan, isfinite
* #21129 Added factorial
* #21128 Finished trigonometric functions
* #21127 Finished the high-priority functions
* #21126 Added remainder of high-priority trigonometric math ops
* **#21125 Added better tests for math ops and unified them**

Differential Revision: [D15563188](https://our.internmc.facebook.com/intern/diff/D15563188)

Chillee committed:
Added better tests for math ops and unified them

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee commented on test/test_jit.py:
```diff
@@ -5952,84 +5952,86 @@ def func(a, b):
+                if debug:
+                    print("in: ", a, b)
+                    print("out: ", resf, resfs)
+                if resf != resfs:
```
I can't use `assertEqual` here because there are some specific conditions (`nan` should be equal, we want to silently ignore errors, etc.)

Actually, lemme document that.

driazati commented on test/test_jit.py:
```diff
@@ -5952,84 +5952,86 @@ def func(a, b):
-            return math.log10(x)
+        inf = float("inf")
+        NaN = float("nan")
+        mx_int = 2147483647
```
we use `int64_t` internally so we should try be `2 ** 31`, `2 ** 32`, and `2 ** 63` (and the negations)

Chillee replied:
I do the first 2 - I'll add the last one.

Chillee replied:
Actually, I don't think it's a good idea. Python has big ints, so it'll invariably have different behavior than what we do near the boundary.

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

driazati commented on test/test_jit.py:
```diff
@@ -5952,84 +5952,86 @@ def func(a, b):
+        NaN = float("nan")
+        mx_int = 2147483647
+        mn_int = -2147483647 - 1
+        float_vals = [inf, NaN, 0.0, 1.0, 2.2, -1.0, -0.0, -2.2, -inf, 1, 0, 2]
```
Test big values for float too

driazati commented on test/test_jit.py:
```diff
@@ -5952,84 +5952,86 @@ def func(a, b):
+            for a, b in vals:
+                resf = None
+                resfs = None
+                try:
```
This flow is hard to follow (and the names could be more descriptive), it seems like it could mostly be replaced with `self.assertEqual`

Chillee replied:
There's a comment about the conditions that `self.assertEqual` doesn't support - is it possible to make it support that? https://github.com/pytorch/pytorch/pull/21125/files#diff-78aab5b4217aeaed920efab0c5f5f5c8R6009

I suppose it might be able to replace the last if statement or so...

driazati replied:
Oh yeah I missed that, the `NaN` thing we shouldn't change so I guess it's not all that useful,  what do you mean by `2. We want to return `nan` for math.sqrt(-5)`, shouldn't it be an exception?

driazati commented on test/test_jit.py:
```diff
@@ -5952,84 +5952,86 @@ def func(a, b):
+                                         .format(func_name=func_name, a=a, b=b, resf=resf, resfs=resfs))
+
+
+        unary_float_ops = ["log", "log1p", "log10", "exp", "sqrt", "gamma", "lgamma", "erf", "erfc", "expm1", "fabs"]
```
We should have something like `[x for x in dir(math) if callable(getattr(math, x))]` so we know we're getting full coverage of the `math` module instead of maintaining big lists like these (though we still need some list to say what the args are for the non-unary ones)

Chillee replied:
I'd prefer to do this after this stack lol - otherwise I'd need to maintain a list of what functions aren't supported that I whittle down for each commit in this stack.

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

ailzhang reviewed (approved):
LGTM!

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head

Chillee committed:
Update on "Added better tests for math ops and unified them"

Added better tests for math ops and unified them

gh-metadata: pytorch pytorch 21125 gh/chillee/3/head
